File: common/src/main/java/com/noxcrew/noxesium/network/serverbound/ServerboundClientInformationPacket.java
package com.noxcrew.noxesium.network.serverbound;

import com.noxcrew.noxesium.network.NoxesiumPackets;
import com.noxcrew.noxesium.network.NoxesiumPayloadType;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;

/**
 * Sent to the server when the client first joins to establish the version of the
 * client being used.
 */
public record ServerboundClientInformationPacket(int protocolVersion, String versionString)
        implements ServerboundNoxesiumPacket {
    public static final StreamCodec<RegistryFriendlyByteBuf, ServerboundClientInformationPacket> STREAM_CODEC =
            CustomPacketPayload.codec(
                    ServerboundClientInformationPacket::write, ServerboundClientInformationPacket::new);

    private ServerboundClientInformationPacket(RegistryFriendlyByteBuf buf) {
        this(buf.readInt(), buf.readUtf());
    }

    private void write(RegistryFriendlyByteBuf buf) {
        buf.writeByte(protocolVersion);
        buf.writeUtf(versionString);
    }

    @Override
    public NoxesiumPayloadType<?> noxesiumType() {
        return NoxesiumPackets.SERVER_CLIENT_INFO;
    }
}

File: paper/src/main/kotlin/com/noxcrew/noxesium/paper/api/network/serverbound/ServerboundClientInformationPacket.kt
package com.noxcrew.noxesium.paper.api.network.serverbound

import com.noxcrew.noxesium.api.protocol.NoxesiumFeature
import com.noxcrew.noxesium.paper.api.network.NoxesiumPackets
import net.minecraft.network.RegistryFriendlyByteBuf
import org.bukkit.entity.Player

/**
 * Sent to the server when the client first joins to establish the version of the
 * client being used.
 */
public data class ServerboundClientInformationPacket(
    public val protocolVersion: Int,
    public val versionString: String,
) : ServerboundNoxesiumPacket(NoxesiumPackets.SERVER_CLIENT_INFO) {

    public constructor(buffer: RegistryFriendlyByteBuf, player: Player, protocolVersion: Int) : this(
        buffer,
        buffer.readVarInt(),
    )

    public constructor(buffer: RegistryFriendlyByteBuf, protocolVersion: Int) : this(
        protocolVersion,
        if (protocolVersion >= NoxesiumFeature.NEW_MCC_FEATURES.minProtocolVersion) buffer.readUtf() else "unknown"
    )
}
